### ISSUE TITLE
Update libraries to use nullable reference types and communicate nullable rules to callers|Generic definitions and nullability – outdated?

### ISSUE THREAD

**Issue by Uzivatel919** [dotnet/docs#24980]
From [Generic definitions and nullability](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-migration-strategies#generic-definitions-and-nullability)

> What it does mean is that you can't use T? in a generic class or method declaration without constraints. For example, Enumerable.FirstOrDefault<TSource>(IEnumerable<TSource>) won't be changed to return T?. You can overcome this limitation by adding either the struct or class constraint.

This whole chapter is perhaps outdated(?).

None of the methods bellow produce any compiler problem.

```csharp
[SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demostration!")]
static void Y<T> ( T [] p )
{
  T? y = Enumerable.FirstOrDefault<T> (p);
}

[SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demostration!")]
static void Y3<T> ( T? [] p )
{
  T? y = Enumerable.FirstOrDefault<T?> (p);
}

[SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demostration!")]
static void Y4<T> ( T? [] p )
{
  T? y = Enumerable.FirstOrDefault (p);
}
```

On the other hand `T?` notation is not recognized in method declaration.

```csharp
[SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demostration!")]
static void Y5<T?> ( T[] p ) where T : struct
//              ↑ Error CS1003  Syntax error, ',' expected
{
  T? y = Enumerable.FirstOrDefault<T> (p);
}

[SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demostration!")]
static void Y6<T?> ( T[] p ) where T : class
//              ↑ Error CS1003  Syntax error, ',' expected
{
  T? y = Enumerable.FirstOrDefault<T> (p);
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: fc516159-d7e1-46ba-7d31-538a0b811491
* Version Independent ID: 99552131-8839-7c22-fcdc-2baa7722e756
* Content: [Update your codebase to use nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-migration-strategies)
* Content Source: [docs/csharp/nullable-migration-strategies.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-migration-strategies.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by Uzivatel919** [comment]
I guess I hit the issue what paragraph is trying to state in.

```csharp
 [SuppressMessage ("Style", "IDE0059:Unnecessary assignment of a value", Justification = "Demonstration.")]
static void Test3 ()
{
  {
    T? Method<T> ( T? t ) => t;

    object? t1  = Method<object> (null);
    object? t2  = Method<object?> (null);
    int? t3     = Method<int> (null); // Error	CS1503	Argument 1: cannot convert from '<null>' to 'int'	
    int? t4     = Method<int?> (null);
  }

  {
    T? Method<T> ( T? t ) where T : struct => t;

    int? t1   = Method<int> (null);
    int? t2   = Method<int?> (null); // Error	CS0453	The type 'int?' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Method<T>(T?)'
  }
  {
    T? Method<T> ( T? t ) where T : class => t;

    object? t1  = Method<object> (null);
    object? t2  = Method<object?> (null); // Warning	CS8634	The type 'object?' cannot be used as type parameter 'T' in the generic type or method 'Method<T>(T?)'. Nullability of type argument 'object?' doesn't match 'class' constraint.
    object? t3  = Method ((object?)null);
  }
}
```

It seems `T?` can be used but there can be _enigmatic_ fail for `struct` to `struct?` conversion. It seems that applying `struct` or `class` constraint brings predictable and flat design option.

**Comment by BillWagner** [comment]
Adding to this sprint. I'll make these edits at the same time as addressing #24257

**Comment by Uzivatel919** [comment]
@BillWagner,

I finally found out what is going there. Likely you are interested. Please see https://softwareengineering.stackexchange.com/questions/429942/what-compiler-logic-is-behind-forcing-value-type-parameter-to-be-not-null-in-g#430260.

**Comment by Uzivatel919** [comment]
Observered behavior is defined at [Unconstrained type parameter annotations|? annotation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/unconstrained-type-parameter-annotations#-annotation)

> If T is substituted with a value type, then T? represents an instance of T.

**Comment by BillWagner** [comment]
Fixed in #26046
